btree/btree.c

```diff
@@ -4075,192 +4223,621 @@ static int  vkvv_space(const struct nd *node)
+	int               total_size     = 1ULL << h->vkvv_shift;
+	void             *start_val_addr = (void*)h + total_size;
+	int               index          = slot->s_idx;
+	uint16_t          vspace         = vkvv_get_vspace();
+	void             *new_addr;
+
+	if (h->vkvv_level > 0) {
+		new_addr = start_val_addr - vspace*(index+1) + INT_OFFSET;
+		*addr    = *(struct segaddr *)new_addr;
+	}
 }
 
 static bool vkvv_isfit(struct slot *slot)
 {
 	/**
 	 * @brief This function will determine if the current key and value
 	 *        can be added to the node.
-	 *        It also checks the validity of the record.
-	 *        Modifications will be required in nt_isfit() function
-	 *        declaration as we need to have the sizes of incoming key and
-	 *        value.
-	 *
-	 *        keysize + valsize <= vkvv_space(slot->s_node)
 	 */
+	m0_bcount_t ksize = m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	m0_bcount_t vsize;
+	uint16_t    dir_entry;
+
+	if (vkvv_level(slot->s_node) == 0) {
+		dir_entry = sizeof(struct dir_rec);
+		vsize     = m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	} else {
+		dir_entry = 0;
+		vsize     = vkvv_get_vspace();
+	}
+	return ksize + vsize + dir_entry <= vkvv_space(slot->s_node);
 }
 
 static void vkvv_done(struct slot *slot, bool modified)
 {
 
 }
 
+static bool vkvv_is_dir_overlap(struct slot *slot)
+{
+	uint16_t          ksize          =
+		slot->s_rec.r_key.k_data.ov_vec.v_count[0];
+	uint16_t          vsize          = slot->s_rec.r_val.ov_vec.v_count[0];
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	int               count          = h->vkvv_used;
+
+	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
+	void             *key_addr       = vkvv_key(slot->s_node, count);
+	void             *val_addr       = vkvv_val(slot->s_node, count);
+	void             *end_dir_addr   = start_dir_addr +
+					   sizeof(struct dir_rec)*(count + 2);
+
+	if (key_addr + ksize > start_dir_addr ||
+	    val_addr - vsize < end_dir_addr)
+		return true;
+	else
+		return false;
+}
+
+static void vkvv_move_dir(struct slot *slot)
+{
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	int               count          = h->vkvv_used;
+	uint16_t          ksize          =
+		slot->s_rec.r_key.k_data.ov_vec.v_count[0];
+	uint16_t          vsize          = slot->s_rec.r_val.ov_vec.v_count[0];
+
+	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
+	uint16_t          dir_size       = sizeof(struct dir_rec)*(count + 2);
+	void             *end_dir_addr   = start_dir_addr + dir_size;
+	void             *key_addr       = vkvv_key(slot->s_node, count);
+	void             *val_addr       = vkvv_val(slot->s_node, count);
+	void             *temp_dir_buffer;
+	uint16_t          diff;
+
+	/* Key space overlaps with directory */
+	if (key_addr + ksize > start_dir_addr) {
+		diff = (key_addr + ksize) - start_dir_addr;
+		temp_dir_buffer = m0_alloc(dir_size);
```

Comment:
Do not use temp buffer to make double copy. Use memmove function call instead it will take care of overlaps. You might want to add that call in a file in lib/ folder so that it becomes available for all the routines.
Please make similar changes to other places where you have done double-copy.

Reply:
Updated